lib/recursive-methods.rb

```diff
@@ -1,49 +1,88 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n <= -1
+    return 1 if n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+
+    return "" if s.length == 0
+    return s if s.length == 1
+
+    return s[s.length-1].chr + reverse(s[0, s.length-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: 0(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return "" if s.length == 0
+    return s if s.length == 1
+
+    return s[s.length-1].chr + reverse_inplace(s[0, s.length-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    if n == 0
+        return 0
+    else
+        return 2 + bunny(n - 1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def nested(s)
```

Comment:
👍 However because you have a slice with each recursive call, this is O(n^2) for both time and space complexity.